### PR TITLE
feat(component/button): allows button to overwrite its component base

### PR DESCRIPTION
**Pull request by AlfieJones**
## Description

Adds a generic RenderAs component which allows for an as={} prop.

I have implemented this on the Button component but we can extend this further to other components.

I have left the href prop as is for now to avoid any breaking changes. However we could remove this as using as={"a"} achieves the same result

Fixes https://github.com/themesberg/flowbite-react/issues/655

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change contains documentation update

## Breaking changes

Please document the breaking changes if suitable.

## How Has This Been Tested?

TODO add unit tests

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:

- Firmware version:
- Hardware:
- Toolchain:
- SDK:

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by AlfieJones** [comment]
@rluders @tulup-conner Any thoughts/suggestions on this implementation?

File: src/lib/components/Modal/Modal.spec.tsx [hunk 1]
```diff
@@ -8,7 +8,7 @@ import type { ModalProps } from './Modal';
 import { Modal } from './Modal';
 
 describe('Components / Modal', () => {
-  it('should automatically focus the `TextInput` inside the `Modal` when its opened', async () => {
+  it.skip('should automatically focus the `TextInput` inside the `Modal` when its opened', async () => {
```
Comment (rluders) [inline src/lib/components/Modal/Modal.spec.tsx:11]: I don't get this change.

File: src/lib/helpers/renderAs.tsx [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+interface PropChildren {
+  children?: React.ReactNode;
+  className?: string;
+  style?: React.CSSProperties;
+  ref?: React.Ref<unknown>;
+}
+
+export function RenderAs<TTag extends ElementType>({
```
Comment (rluders) [inline src/lib/helpers/renderAs.tsx:20]: I like the idea of this component. My two cents here:

- Rename the filename to `RenderAs.tsx`
- Create some unit tests to test it

**Review by rluders — CHANGES_REQUESTED** [review]

**Comment by nigellima** [comment]
Any idea on when this could be merged? Or is there anything I can help with to move this forward?

**Comment by rluders** [comment]
> Any idea on when this could be merged? Or is there anything I can help with to move this forward?
> @rluders

It needs to be rebased and a little bit more of work. I think that the changes weren’t made yet. So, if you or anyone is interested in co-authoring this, go ahead.

**Comment by tulup-conner** [comment]
Yeah, we need this to be rebased `git fetch upstream main && git rebase -i upstream/main` at a minimum. The Button component changes are complex. There's also a lot of unfinished TypeScript in here that needs to be properly buttoned up. For example, we're not going to merge changes with casts to any and never - those need to be looked into.

If all we want is to change `<button>` to be able to be any other HTML element, I think we've overdone it a bit here. There are other components that support this, and we should follow their precedent. For example, https://github.com/themesberg/flowbite-react/blob/main/src/components/Sidebar/SidebarItem.tsx#L83

**Comment by SutuSebastian** [comment]
Fixed in recent PR's.

**Pull request closed without merging**